Add JavaScript API Reference

romaricpascal opened the pull request:
Pulls the content from the documentation draft into a new JavaScript API Reference page and formats it. 

The page is brokent down by components, then option. Complex types, like `AccordionTranslations` or `CharacterCountTranslations` that were present in the actual JSDoc have been dropped in favour of a more straightforward `object` and the list of properties straight after.

The defaults have been formatted as a code block. This is not only consistent with how they're displayed in the Sass reference, but also allows the same display whether they're for a scalar value or an `Object`/`Array` (as well as providing a bit of break while going through the page).

Closes #215

colinrotherham commented on source/configure-components-with-javascript/index.html.md.erb:
```diff
@@ -0,0 +1,94 @@
+  characterCount: {
+    maxlength: 500, // Non namespaced
+    i18n: { // i18n namespace
+      charactersAtLimit: “No characters left”
```
Automatic smart quote conversion 😮

```suggestion
      charactersAtLimit: "No characters left"
```

colinrotherham commented on source/configure-components-with-javascript/index.html.md.erb:
```diff
@@ -0,0 +1,94 @@
+
+## Passing configuration using data attributes in HTML
+
+If you're using HTML, you can pass configuration by adding data attributes to the component's outermost element (the element that has the `data-module` attribute). This is how our Nunjucks macros forward the configuration to the JavaScript components in the browser. Data attributes use kebab-case.
```
```suggestion
If you’re using HTML, you can pass configuration by adding data attributes to the component’s outermost element (the element that has the `data-module` attribute). This is how our Nunjucks macros forward the configuration to the JavaScript components in the browser. Data attributes use kebab-case.
```
```suggestion
If you're using HTML, you can pass configuration by adding data attributes to the component's outermost element (the element that has the `data-module` attribute). This is how our Nunjucks macros forward the configuration to the JavaScript components in the browser. Data attributes use kebab-case.
```

romaricpascal replied:
Middleman (or Kramdown, or something in between) takes care of making the quotes smart inside our content, so we can author them as `'` and they appear OK [on the generated page](https://deploy-preview-219--govuk-frontend-docs-preview.netlify.app/localise-govuk-frontend/).

colinrotherham replied:
@romaricpascal But we're mixing styles? Straight single quote here, curly ones on lines 29, 45, 49 etc?

That alright?

romaricpascal replied:
Oh, I see. Seems the other files are quite lenient on having those consistent, but I'll tidy it up in these new pieces of documentation. Wonder if that's something we can lint, maybe with `remark-lint` :D

colinrotherham commented on source/localise-govuk-frontend/index.html.md.erb:
````diff
@@ -0,0 +1,137 @@
+With data attributes
+
+```html
+<div data-module=”character-count” 
````
I won't flag them all, but this block's been formatted with smart quotes too 🙈 

```suggestion
<div data-module="character-count" 
```

colinrotherham commented on source/localise-govuk-frontend/index.html.md.erb:
```diff
@@ -0,0 +1,137 @@
+    },
+    charactersOverLimitText: {
+      one: 'One character too many'
+      // The `other` key for `characterOverLimitText` will be the default one
```
We have `other` as mandatory in our JSDoc so leave it out here

romaricpascal replied:
The JSDoc had an error for that one as you may want to pass an object without the `other` key when overriding only one of our English defaults, like in the example.

colinrotherham commented on source/localise-govuk-frontend/index.html.md.erb:
```diff
@@ -0,0 +1,137 @@
+
+## Changing the text in the HTML for a component
+
+You can use the Nunjucks macro options tables published on the GOV.UK Design System site to find the option you need to change the text. The conventions for these options are:
```
```suggestion
You can use the Nunjucks macro options tables published on the GOV.UK Design System website to find the option you need to change the text. The conventions for these options are:
```

romaricpascal commented on source/configure-components-with-javascript/index.html.md.erb:
```diff
@@ -0,0 +1,94 @@
+
+## Setting Nunjucks macro options
+
+If you're using the Nunjucks macros, you will find all of the configuration options for a component published on the GOV.UK Design System site.
```
Updated, alongside another occurence in the localisation docs.

36degrees commented on source/localise-govuk-frontend/index.html.md.erb:
````diff
@@ -0,0 +1,137 @@
+JavaScript uses nested objects. For example:
+
+```javascript
+{
+  i18n: {
+  camelCasedName: /* Text or Object with plural forms */
+  }
+}
````
Minor indentation:

```suggestion
{
  i18n: {
    camelCasedName: /* Text or Object with plural forms */
  }
}
```

36degrees commented on source/localise-govuk-frontend/index.html.md.erb:
````diff
@@ -0,0 +1,137 @@
+With Nunjucks
+
+```njk
+{{ govukCharacterCount({
+  charactersUnderLimitText: {
+    other: '%{count} characters to go'
+    one: 'One character to go'
+  },
+  charactersOverLimitText: {
+    one: 'One character too many'
+  }
+})
````
Couple of syntax errors:

```suggestion
{{ govukCharacterCount({
  charactersUnderLimitText: {
    other: '%{count} characters to go',
    one: 'One character to go'
  },
  charactersOverLimitText: {
    one: 'One character too many'
  }
}) }}
```

romaricpascal replied:
I was seeing these examples more illustrative of the localisation options than copy-pasteable to get a CharacterCount up and running (which I'd expect more on the Character Count page). Looking at the docs and the code, we're also missing the `label` (which won't be graceful if missing either) and the `name`: 

```suggestion
{{ govukCharacterCount({
  id: 'details',
  name: 'details',
  label: {
    html: 'Can you provide more details?'
  },
  maxlength: 1000,
  charactersUnderLimitText: {
    other: '%{count} characters to go',
    one: 'One character to go'
  },
  charactersOverLimitText: {
    one: 'One character too many'
  }
}) }}
```

36degrees commented on source/localise-govuk-frontend/index.html.md.erb:
````diff
@@ -0,0 +1,137 @@
+With JavaScript
+
+```js
+new CharacterCount($element, {
+  i18n: {
+    charactersUnderLimit: {
+      other: '%{count} characters to go'
+      one: 'One character to go'
+    },
+    charactersOverLimit: {
+      one: 'One character too many'
+      // The `other` key for `characterOverLimit` will be our default
+    }
+  }
+}).init()
````
```suggestion
new CharacterCount($element, {
  i18n: {
    charactersUnderLimit: {
      other: '%{count} characters to go',
      one: 'One character to go'
    },
    charactersOverLimit: {
      one: 'One character too many'
      // The `other` key for `characterOverLimit` will be our default
    }
  }
}).init()
```

36degrees commented on source/configure-components-with-javascript/index.html.md.erb:
```diff
@@ -0,0 +1,94 @@
+- setting Nunjucks macro options
+- using data attributes in the HTML
+- passing a JavaScript object when creating an instance of a component
+- using the initAll function
```
Should this be backticked?

```suggestion
- using the `initAll` function
```

romaricpascal replied:
Oh, indeed. Cheers!

36degrees reviewed (commented):
Generally this is looking good – just a few minor things that need correcting.

romaricpascal commented:
- @36degrees Code blocks have been updated
- @colinrotherham curly quotes have been removed from this PR so it's all straight quotes consistently (in these files at least)

Feels like this is now in a good place, waiting for the release 🎉

36degrees reviewed (commented):
I don't think this should block the release, but the syntax highlighting within the JS API Reference is a little off, as the JS isn't really 'complete':

![Screenshot 2022-11-10 at 13 26 29](https://user-images.githubusercontent.com/121939/201104171-bcaf1c06-09bc-4f20-89d3-13df71f9c629.png)

Marking them up as JSON improves things slightly…

![Screenshot 2022-11-10 at 13 26 34](https://user-images.githubusercontent.com/121939/201104175-d91fde59-d0fe-4a69-93c4-bc5909405def.png)

But seem to get the best result if we switch to using double quotes for strings:

![Screenshot 2022-11-10 at 13 27 51](https://user-images.githubusercontent.com/121939/201104176-3332e2e4-636d-4974-8e74-c8db01f81a58.png)

romaricpascal commented:
Updated the default values to be marked as JSON for now (and switched to double quotes, which makes it proper JSON). That'll be something to keep in mind when we pull the defaults straight from JSDoc (and if we start having arrow functions as defaults at some points).
